src/inetstack/protocols/layer4/tcp/established/receiver.rs

```diff
@@ -112,31 +116,43 @@ impl Receiver {
     }
 
     // Block until some data is received, up to an optional size.
-    pub async fn pop(&mut self, size: Option<usize>) -> Result<DemiBuffer, Fail> {
-        debug!("waiting on pop {:?}", size);
-        let buf: DemiBuffer = if let Some(size) = size {
-            let mut buf: DemiBuffer = self.pop_queue.pop(None).await?;
-            // Split the buffer if it's too big.
-            if buf.len() > size {
-                buf.split_front(size)?
-            } else {
-                buf
+    pub async fn pop(
+        &mut self,
+        mut size: Option<usize>,
+    ) -> Result<ArrayVec<DemiBuffer, MAX_BATCH_SIZE_NUM_PACKETS>, Fail> {
+        let mut bufs: ArrayVec<DemiBuffer, MAX_BATCH_SIZE_NUM_PACKETS> = ArrayVec::new();
+        let mut buf: DemiBuffer = self.pop_queue.pop(None).await?;
+        loop {
+            if let Some(size) = size.as_mut() {
+                if buf.len() > *size {
+                    let remaining_buf: DemiBuffer = buf.split_front(*size)?;
+                    self.pop_queue.push_front(remaining_buf);
+                }
+                *size = *size - buf.len();
             }
-        } else {
-            self.pop_queue.pop(None).await?
-        };
+            match buf.len() {
+                len if len > 0 => {
+                    self.reader_next_seq_no = self.reader_next_seq_no + SeqNumber::from(buf.len() as u32);
+                },
+                _ => {
+                    debug!("found FIN");
+                    self.reader_next_seq_no = self.reader_next_seq_no + 1.into();
+                    bufs.push(buf);
 
-        match buf.len() {
-            len if len > 0 => {
-                self.reader_next_seq_no = self.reader_next_seq_no + SeqNumber::from(buf.len() as u32);
-            },
-            _ => {
-                debug!("found FIN");
-                self.reader_next_seq_no = self.reader_next_seq_no + 1.into();
-            },
+                    break;
+                },
+            }
+            bufs.push(buf);
+            match size {
+                Some(size) if size == 0 => break,
```

Comment:
You can simply use this representation instead to avoid specifying the condition
```suggestion
                Some(0) => break,
```